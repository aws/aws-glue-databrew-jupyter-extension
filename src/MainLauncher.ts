import { Widget as PhosphorWidget } from "@phosphor/widgets";
import { Widget as LuminoWidget } from "@lumino/widgets";
import { AWSRegion } from "./types";
import { isCnPartition } from "./utils";
export class PhosphorMainLauncher extends PhosphorWidget {
  /**
   * The image element associated with the widget.
   */
  consoleRoot: HTMLElement;
  /**
   * A path to css assets
   */
  cssPath: string;
}

export class LuminoMainLauncher extends LuminoWidget {
  /**
   * The image element associated with the widget.
   */
  consoleRoot: HTMLElement;
  /**
   * A path to css assets
   */
  cssPath: string;
}

export class MainLauncher {
  static create(version: string, baseUrl: string, cssPath: string, region: AWSRegion): PhosphorMainLauncher | LuminoMainLauncher {
    const majorVersion = Number(version.split(".")[0]);
    const widget = majorVersion === 1 ? new PhosphorMainLauncher() : new LuminoMainLauncher();

    const awsText = isCnPartition(region) ? "Amazon" : "AWS";
    widget.cssPath = cssPath;
    widget.id = "aws_glue_databrew_jupyter";
    widget.title.label = `${awsText} Glue DataBrew`;
    widget.title.closable = true;

    widget.consoleRoot = document.createElement("html");
    widget.consoleRoot.setAttribute("style", "height: 100%");
    widget.node.appendChild(widget.consoleRoot);

    widget.consoleRoot.insertAdjacentHTML("beforeend",
      `
        <head>
          <meta charset="UTF-8">
          <meta name="awsc-lang" content="en">
          <meta id="jupyter-server-base-url" content="${baseUrl}">
          <meta id="jupyterlab-version" content="${version}">
          <meta name="aws-glue-databrew-jupyter" content="true">
          <meta id="aws-glue-databrew-jupyter-region" content="${region || ""}">
          <title>AWS</title>
          <link rel="stylesheet" href="${cssPath}">
          <style>
            .loader {
              display: inline-block;
              border: 5px solid #f3f3f3;
              border-radius: 50%;
              border-top: 5px solid #3498db;
              width: 40px;
              height: 40px;
              margin-right: 12px;
              -webkit-animation: spin 2s linear infinite;
              animation: spin 2s linear infinite;
            }

            @-webkit-keyframes spin {
              0% { -webkit-transform: rotate(0deg); }
              100% { -webkit-transform: rotate(360deg); }
            }

            @keyframes spin {
              0% { transform: rotate(0deg); }
              100% { transform: rotate(360deg); }
            }

            .loader-container {
              font-size: 60px;
              font-weight: 100;
              color: #879196;
              height: 100%;
              width: 100%;
              justify-content: center;
              display: flex;
              align-items: center;
            }
          </style>
        </head>
        <body style="height: 100%">
          <div id="app" style="height: 100%">
            <div class="loader-container">
              <div>
                <div class="loader"></div><span>Launching ${awsText} Glue DataBrew</span>
              </div>
            </div>
          </div>
        </body>
    `
    );

    return widget;
  }
}
