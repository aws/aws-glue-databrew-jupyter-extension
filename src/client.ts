import { AWSConfig } from "./types";

export const getAWSConfig = async (basePath: string): Promise<AWSConfig> => {
  try {
    const config = await get<AWSConfig>(`${basePath}awsproxy/awsconfig`);
    return {
      region: config.region,
    };
  } catch {
    const config = await post<AWSConfig>(`${basePath}api/getInstanceConfig`);
    return {
      region: config.region,
    };
  }
};

export const get = <T>(path: string): Promise<T> => {
  const options = { method: "get" };
  return request(new Request(path, options));
};

export const post = <T>(path: string): Promise<T> => {
  const options = { method: "post" };
  return request(new Request(path, options));
};

export const request = async <T>(req: Request): Promise<T> => {
  const res = await fetch(req);
  const text = await res.text();
  const data = text && parse<T | Error>(text);
  if (!res.ok) {
    const error = data as Error;
    throw new Error((error && error.message) || res.statusText);
  }
  return data as T;
};

export const parse = <T>(text: string): T => {
  try {
    return JSON.parse(text);
  } catch {
    return {} as T;
  }
};
