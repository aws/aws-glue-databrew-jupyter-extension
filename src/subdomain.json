{"subdomain": "dc69wn9dwut4o"}
