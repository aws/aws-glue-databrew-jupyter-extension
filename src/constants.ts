import { subdomain } from "./subdomain.json";

export const GLUE_DATABREW_RENDER = "gluedatabrew:render";
export const CN_REGION_HOST = "https://elixir-console-assets-prod-bjs.s3.cn-north-1.amazonaws.com.cn"
export const CLOUDFRONT_HOST = `https://${subdomain}.cloudfront.net`;
export const jsFileName = "main.js";
export const styleFileName = "styles.css";
