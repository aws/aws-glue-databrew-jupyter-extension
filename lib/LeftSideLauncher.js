import { Widget as PhosphorWidget } from "@phosphor/widgets";
import { Widget as LuminoWidget } from "@lumino/widgets";
import { GLUE_DATABREW_RENDER } from "./constants";
export class PhosphorLeftSideLauncher extends PhosphorWidget {
    handleLaunchButtonClicked() {
        this.commands.execute(GLUE_DATABREW_RENDER);
        return;
    }
}
export class LuminoLeftSideLauncher extends LuminoWidget {
    handleLaunchButtonClicked() {
        this.commands.execute(GLUE_DATABREW_RENDER);
        return;
    }
}
export class LeftSideLauncher {
    static create(version, commands) {
        const majorVersion = Number(version.split(".")[0]);
        const widget = majorVersion === 1 ? new PhosphorLeftSideLauncher() : new LuminoLeftSideLauncher();
        widget.commands = commands;
        const launchButton = document.createElement("div");
        const header = document.createElement("header");
        header.className = "aws_glue_databrew_header";
        header.textContent = "AWS Glue DataBrew";
        launchButton.title = "Welcome to AWS Glue DataBrew";
        launchButton.id = "aws_glue_databrew_launch_button";
        launchButton.textContent = "Launch AWS Glue DataBrew";
        launchButton.onclick = widget.handleLaunchButtonClicked.bind(widget);
        widget.node.appendChild(header);
        widget.node.appendChild(launchButton);
        widget.id = "aws_glue_databrew_jupyter_left_side_launcher";
        widget.title.iconClass = "jp-databrew-logo jp-SideBar-tabIcon";
        widget.title.caption = "AWS Glue DataBrew";
        return widget;
    }
}
