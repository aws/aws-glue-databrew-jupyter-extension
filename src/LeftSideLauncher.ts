import { Widget as PhosphorWidget } from "@phosphor/widgets";
import { CommandRegistry as PhosphorCommandRegistry } from "@phosphor/commands";
import { Widget as LuminoWidget } from "@lumino/widgets";
import { CommandRegistry as LuminoCommandRegistry } from "@lumino/commands";
import { GLUE_DATABREW_RENDER } from "./constants";

export class PhosphorLeftSideLauncher extends PhosphorWidget {
  /**
   * Command Registry
   */
  commands: PhosphorCommandRegistry;
  handleLaunchButtonClicked(): undefined {
    this.commands.execute(GLUE_DATABREW_RENDER);
    return;
  }
}

export class LuminoLeftSideLauncher extends LuminoWidget {
  /**
   * Command Registry
   */
  commands: LuminoCommandRegistry;
  handleLaunchButtonClicked(): undefined {
    this.commands.execute(GLUE_DATABREW_RENDER);
    return;
  }
}

export class LeftSideLauncher {
  static create(version: string, commands: PhosphorCommandRegistry | LuminoCommandRegistry): PhosphorLeftSideLauncher | LuminoLeftSideLauncher {
    const majorVersion = Number(version.split(".")[0]);
    const widget = majorVersion === 1 ? new PhosphorLeftSideLauncher() : new LuminoLeftSideLauncher();
    widget.commands = commands;

    const launchButton = document.createElement("div");
    const header = document.createElement("header");
    header.className = "aws_glue_databrew_header";
    header.textContent = "AWS Glue DataBrew";

    launchButton.title = "Welcome to AWS Glue DataBrew";
    launchButton.id = "aws_glue_databrew_launch_button";
    launchButton.textContent = "Launch AWS Glue DataBrew";

    launchButton.onclick = widget.handleLaunchButtonClicked.bind(widget);

    widget.node.appendChild(header);
    widget.node.appendChild(launchButton);
    widget.id = "aws_glue_databrew_jupyter_left_side_launcher";
    widget.title.iconClass = "jp-databrew-logo jp-SideBar-tabIcon";
    widget.title.caption = "AWS Glue DataBrew";

    return widget;
  }
}
