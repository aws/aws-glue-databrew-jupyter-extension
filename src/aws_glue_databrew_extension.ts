import {
  JupyterFrontEnd,
  JupyterFrontEndPlugin,
  ILayoutRestorer,
} from "@jupyterlab/application";
import { ICommandPalette } from "@jupyterlab/apputils";
import { ILauncher } from "@jupyterlab/launcher";
import { Widget } from "@lumino/widgets";
import { getAWSConfig } from "./client";
import { GLUE_DATABREW_RENDER } from "./constants";
import { LeftSideLauncher } from "./LeftSideLauncher";
import { MainLauncher } from "./MainLauncher";
import { isCnPartition } from "./utils";

const getAppVersion = (app: JupyterFrontEnd) => app.version;
const getBaseUrl = (app: JupyterFrontEnd) =>
  app.serviceManager.serverSettings.baseUrl;

/**
 * Initialize the console widget extension
 */
export const initiateExtension = (
  getPaths: (region: string) => Promise<string[]>
): JupyterFrontEndPlugin<void> => {
  const activate = async (
    app: JupyterFrontEnd,
    palette: ICommandPalette,
    restorer: ILayoutRestorer,
    launcher: ILauncher
  ) => {
    const version = getAppVersion(app);
    const baseUrl = getBaseUrl(app);
    const url = new URL(baseUrl);
    const { region } = await getAWSConfig(url.pathname);
    const [jsPath, cssPath] = await getPaths(region);
    const awsText = isCnPartition(region) ? "Amazon" : "AWS";
    const consoleWidget = MainLauncher.create(
      version,
      baseUrl,
      cssPath,
      region
    );

    app.commands.addCommand(GLUE_DATABREW_RENDER, {
      label: `Launch ${awsText} Glue DataBrew`,
      icon: "jp-databrew-logo",
      execute: () => {
        if (!consoleWidget.isAttached) {
          app.shell.add(consoleWidget as Widget, "main");
          const script = document.createElement("script");

          script.setAttribute("src", jsPath);
          consoleWidget.consoleRoot.appendChild(script);
        }
        app.shell.activateById(consoleWidget.id);
      },
    });

    const launcherWidget = LeftSideLauncher.create(version, app.commands);

    restorer.add(launcherWidget as Widget, launcherWidget.id);
    app.shell.add(launcherWidget as Widget, "left");

    // Add the command to the palette.
    palette.addItem({ command: GLUE_DATABREW_RENDER, category: "Launcher" });
    if (launcher) {
      const launcher_item: ILauncher.IItemOptions = {
        command: GLUE_DATABREW_RENDER,
        args: {
          newBrowserTab: true,
          title: "Launch Databrew",
          id: "databrew-launcher",
        },
        category: "Other",
      };

      launcher.add(launcher_item);
    }
  };

  const extension: JupyterFrontEndPlugin<void> = {
    id: "aws_glue_databrew_jupyter",
    autoStart: true,
    requires: [ICommandPalette, ILayoutRestorer, ILauncher],
    activate: activate,
  };

  return extension;
};
