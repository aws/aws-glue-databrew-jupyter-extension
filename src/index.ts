import { initiateExtension } from "./aws_glue_databrew_extension";
import {
  getContentPrefixUrl,
  getPrefixUrl,
} from "./utils";
import { jsFileName, styleFileName } from "./constants";

export const getPaths = async (region: string): Promise<string[]> => {
  const prefix = await getPrefix(region);
  const contentPrefixUrl = `${getContentPrefixUrl(region)}/${prefix}`;
  return [
    `${contentPrefixUrl}/${jsFileName}`,
    `${contentPrefixUrl}/${styleFileName}`,
  ];
};

export const getPrefix = async (region: string): Promise<string> => {
  const prefixUrl = getPrefixUrl(region);
  const req = new Request(prefixUrl);
  const response = await fetch(req);
  const json = await response.json();
  return json && json.prefix.toString();
};

const extension = initiateExtension(getPaths);

export default extension;
