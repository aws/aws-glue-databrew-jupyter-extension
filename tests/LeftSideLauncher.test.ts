import { CommandRegistry } from "@phosphor/commands";
import { LeftSideLauncher } from "../src/LeftSideLauncher";

describe("LeftSideLauncher", () => {
  describe("create", () => {
    test.each([["1.0.0"], ["2.0.0"]])("should create launcher v%s", (version) => {
      const execute = jest.fn(() => undefined);
      const command: Partial<CommandRegistry> = { execute };
      const widget = LeftSideLauncher.create(version, command as CommandRegistry);
      expect(widget.id).toBe("aws_glue_databrew_jupyter_left_side_launcher");
      expect(widget.title.iconClass).toBe(
        "jp-databrew-logo jp-SideBar-tabIcon"
      );
      expect(widget.title.caption).toBe("AWS Glue DataBrew");
      expect(widget.node).toMatchSnapshot();
      expect(widget.handleLaunchButtonClicked()).toBe(undefined);
      expect(execute).toHaveBeenCalledTimes(1);
      expect(execute).toHaveBeenCalledWith("gluedatabrew:render");
    });
  });
});
