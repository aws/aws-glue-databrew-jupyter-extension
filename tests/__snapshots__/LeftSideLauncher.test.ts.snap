// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LeftSideLauncher create should create launcher v1.0.0 1`] = `
<div
  class="p-Widget"
  id="aws_glue_databrew_jupyter_left_side_launcher"
>
  <header
    class="aws_glue_databrew_header"
  >
    AWS Glue DataBrew
  </header>
  <div
    id="aws_glue_databrew_launch_button"
    title="Welcome to AWS Glue DataBrew"
  >
    Launch AWS Glue DataBrew
  </div>
</div>
`;

exports[`LeftSideLauncher create should create launcher v2.0.0 1`] = `
<div
  class="lm-Widget p-Widget"
  id="aws_glue_databrew_jupyter_left_side_launcher"
>
  <header
    class="aws_glue_databrew_header"
  >
    AWS Glue DataBrew
  </header>
  <div
    id="aws_glue_databrew_launch_button"
    title="Welcome to AWS Glue DataBrew"
  >
    Launch AWS Glue DataBrew
  </div>
</div>
`;
